### PR TITLE
add export report

### PR DESCRIPTION
**Pull request by zhangchiqing**
adding the export report back which [was removed by this PR](https://github.com/onflow/flow-go/pull/4401)

### CODE REVIEW

File: cmd/util/ledger/reporters/export_reporter.go [hunk 1]
```diff
@@ -24,18 +24,15 @@ type ExportReport struct {
 // ExportReporter writes data that can be leveraged outside of extraction
 type ExportReporter struct {
 	logger                   zerolog.Logger
-	chain                    flow.Chain
 	getBeforeMigrationSCFunc GetStateCommitmentFunc
 }
 
 func NewExportReporter(
 	logger zerolog.Logger,
-	chain flow.Chain,
```
Comment (zhangchiqing) [inline cmd/util/ledger/reporters/export_reporter.go:33]: This reporter doesn't use this data

File: ledger/complete/ledger.go [hunk 2]
```diff
@@ -474,10 +478,3 @@ func (l *Ledger) keepOnlyOneTrie(state ledger.State) error {
 	defer l.wal.UnpauseRecord()
 	return l.forest.PurgeCacheExcept(ledger.RootHash(state))
 }
-
-func writeStatusFile(fileName string, e error) error {
```
Comment (zhangchiqing) [inline ledger/complete/ledger.go:478]: Moving it to execution_state_extract.go instead.

File: ledger/complete/ledger.go [hunk 3]
```diff
@@ -335,7 +334,36 @@ func (l *Ledger) ExportCheckpointAt(
 	targetPathFinderVersion uint8,
 	outputDir, outputFile string,
 ) (ledger.State, error) {
+	newTrie, err := l.MigrateAt(state, migrations, targetPathFinderVersion)
```
Comment (zhangchiqing) [inline ledger/complete/ledger.go:337]: extract the logic into `MigrateAt` which will be reused by state extraction

File: cmd/util/cmd/execution-state-extract/execution_state_extract.go [hunk 4]
```diff
@@ -80,10 +85,31 @@ func extractExecutionState(
...
+	// create reporter
+	reporter := reporters.NewExportReporter(log,
+		func() flow.StateCommitment { return targetHash },
+	)
+
+	newMigratedState := ledger.State(newTrie.RootHash())
+	err = reporter.Report(nil, newMigratedState)
+	if err != nil {
+		log.Error().Err(err).Msgf("can not generate report for migrated state: %v", newMigratedState)
+	}
```
Comment (zhangchiqing) [inline cmd/util/cmd/execution-state-extract/execution_state_extract.go:108]: adding back the reporter here

**Comment by sjonpaulbrown** [comment]
I have validated that the report is properly generated during state extraction and has the correct format.

File: ledger/complete/ledger.go [hunk 5]
```diff
@@ -335,7 +334,36 @@ func (l *Ledger) ExportCheckpointAt(
...
+	if err != nil {
+		return ledger.State(hash.DummyHash), fmt.Errorf("fail to migrate: %w", err)
+	}
+
+	statecommitment := ledger.State(newTrie.RootHash())
```
[the remark this answers is not in the document]
Comment (zhangchiqing) [reply ledger/complete/ledger.go]: I wanted to deprecate ledger.ExportCheckpointAt, the caller could just use `MigrateAt` and `wal.StoreCheckpointV6Concurrently` instead.

File: ledger/complete/ledger.go [hunk 6]
```diff
@@ -329,13 +327,32 @@ func (l *Ledger) Checkpointer() (*realWAL.Checkpointer, error) {
 }
 
 // ExportCheckpointAt exports a checkpoint at specific state commitment after applying migrations and returns the new state (after migration) and any errors
+// deprecated
```
Comment (pattyshack) [inline ledger/complete/ledger.go:330]: just delete the code instead leaving it around.  nothing else is using it ...

**Review by janezpodhostnik — APPROVED** [review]

**Pull request merged**